doc/mcxq.1
.\" Copyright (c) 2009 Stijn van Dongen
.TH "mcx q" 1 "18 Sep 2009" "mcx q 1\&.008, 09-261" "USER COMMANDS "
.po 2m
.de ZI
.\" Zoem Indent/Itemize macro I.
.br
'in +\\$1
.nr xa 0
.nr xa -\\$1
.nr xb \\$1
.nr xb -\\w'\\$2'
\h'|\\n(xau'\\$2\h'\\n(xbu'\\
..
.de ZJ
.br
.\" Zoem Indent/Itemize macro II.
'in +\\$1
'in +\\$2
.nr xa 0
.nr xa -\\$2
.nr xa -\\w'\\$3'
.nr xb \\$2
\h'|\\n(xau'\\$3\h'\\n(xbu'\\
..
.if n .ll -2m
.am SH
.ie n .in 4m
.el .in 8m
..
.SH NAME
mcxq \- compute simple graph statistics
.SH SYNOPSIS

mcxq is not in actual fact a program\&. This manual
page documents the behaviour and options of the mcx program when
invoked in mode \fIq\fP\&. The options \fB-h\fP, \fB--apropos\fP,
\fB--version\fP, \fB-set\fP, \fB--nop\fP, \fB-progress\fP\ \&\fI<num>\fP
are accessible
in all \fBmcx\fP modes\&. They are described
in the \fBmcx\fP manual page\&.

\fBmcxq\fP
\fB[-imx\fP <fname> (\fIspecify matrix input\fP)\fB]\fP
\fB[-o\fP <fname> (\fIoutput file name\fP)\fB]\fP
\fB[--edge\fP (\fIlist simple edge statistics\fP)\fB]\fP
\fB[--edge-list\fP (\fIlist edge statistics per node\fP)\fB]\fP
\fB[--dim\fP (\fIonly list domain sizes\fP)\fB]\fP
\fB[-h\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--apropos\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--version\fP (\fIprint version, exit\fP)\fB]\fP
.SH DESCRIPTION

\fBmcxq\fP computes a number of simple statistics derived from the input
graph\&. Among these are the number of arcs, the node neighbour count
average, standard deviation, and median, and the edge weight
average, standard deviation, and median\&.

When instructed to, it can also compute other quantities related
to the input graph\&. See below\&.
.SH OPTIONS

.ZI 2m "\fB-imx\fP <fname> (\fIinput matrix\fP)"
\&
.br
The file name for input that is in mcl native matrix format\&.
.in -2m

.ZI 2m "\fB-o\fP <fname> (\fIoutput file name\fP)"
\&
.br
Set the name of the file where output should be written to\&.
.in -2m

.ZI 2m "\fB--edge\fP (\fIcompute simple statistics\fP)"
\&
.br
This is the default setting\&. If any other option (e\&.g\&. \fB--dim\fP)
is used, \fB--edge\fP has to be specified if this output mode is to
be activated in parallel\&.
.in -2m

.ZI 2m "\fB--edge-list\fP (\fIcompute simple statistics, node-wisel\fP)"
\&
.br
This outputs for each node its neighbour count and the average
weight of the edges connecting that node with those neighbours\&.
.in -2m

.ZI 2m "\fB--dim\fP (\fIonly list domain sizes\fP)"
\&
.br
This outputs the format (native interchange or native binary)
and the dimensions of the encoded matrix\&.
.in -2m
.SH SEE ALSO

\fBmcxio(5)\fP,
and \fBmclfamily(7)\fP for an overview of all the documentation
and the utilities in the mcl family\&.


NEWS
$Id: NEWS,v 1.11 2003/11/29 21:30:05 flux Exp $

BLAST users may consider using the blast module. Add --enable-blast to
./configure and read the manual pages of mcxdeblast and mcxassemble.  and
optionally mclblastline. The blast module does not conflict with the tribe
module; you can install both of them simultaneously.

Read ChangeLog.


src/shmx/mcxq.c
/*   (C) Copyright 2006, 2007, 2008, 2009 Stijn van Dongen
 *
 * This file is part of MCL.  You can redistribute and/or modify MCL under the
 * terms of the GNU General Public License; either version 3 of the License or
 * (at your option) any later version.  You should have received a copy of the
 * GPL along with MCL, in the file COPYING.
*/


#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <limits.h>
#include <string.h>
#include <ctype.h>
#include <signal.h>
#include <time.h>

#include "mcx.h"

#include "util/types.h"
#include "util/ding.h"
#include "util/ting.h"
#include "util/io.h"
#include "util/err.h"
#include "util/opt.h"
#include "util/compile.h"

#include "impala/matrix.h"
#include "impala/io.h"


int valcmp
(  const void*             i1
,  const void*             i2
)
   {  return *((pval*)i1) > *((pval*)i2) ? 1 : *((pval*)i1) < *((pval*)i2) ? -1 : 0
;  }


enum
{  MY_OPT_IMX = MCX_DISP_UNUSED
,  MY_OPT_DIMENSION
,  MY_OPT_STATISTICS
,  MY_OPT_STATISTICS_LIST
,  MY_OPT_FOUT
}  ;

mcxOptAnchor qOptions[] =
{  {  "-imx"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_IMX
   ,  "<fname>"
   ,  "specify input matrix/graph"
   }
,  {  "--dim"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_DIMENSION
   ,  NULL
   ,  "get matrix dimensions"
   }
,  {  "-o"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_FOUT
   ,  "<fname>"
   ,  "output file name"
   }
,  {  "--edge"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_STATISTICS
   ,  NULL
   ,  "get simple edge weight statistics"
   }
,  {  "--edge-list"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_STATISTICS_LIST
   ,  NULL
   ,  "get simple edge weight statistics for each node separately"
   }
,  {  NULL ,  0 ,  0,  NULL, NULL}
}  ;


static  mcxbool get_dim    =     -1;
static  mcxbool get_stat   =     -1;
static  mcxbool get_stat_list =  -1;
static  mcxIO*  xfout      =     (void*) -1;
static  mcxIO* xfmx_g      =   (void*) -1;


static mcxstatus qInit
(  void
)
   {  get_dim  =  FALSE
   ;  get_stat =  FALSE
   ;  get_stat_list =  FALSE
   ;  xfout    =  mcxIOnew("-", "w")
   ;  xfmx_g   =  mcxIOnew("-", "r")
   ;  return STATUS_OK
;  }



static mcxstatus qArgHandle
(  int optid
,  const char* val
)
   {  switch(optid)
      {  case MY_OPT_IMX
      :  mcxIOnewName(xfmx_g, val)
      ;  break
      ;

         case MY_OPT_DIMENSION
      :  get_dim = TRUE
      ;  break
      ;

         case MY_OPT_STATISTICS_LIST
      :  get_stat_list = TRUE
      ;  break
      ;

         case MY_OPT_FOUT
      :  mcxIOnewName(xfout, val)
      ;  break
      ;

         case MY_OPT_STATISTICS
      :  get_stat = TRUE
      ;  break
      ;

         default
      :  mcxExit(1) 
      ;
   ;  }
      return STATUS_OK
;  }


static mcxstatus qMain
(  int          argc_unused      cpl__unused
,  const char*  argv_unused[]    cpl__unused
)
   {  if (get_dim + get_stat + get_stat_list == 0)
      get_stat = TRUE

   ;  mcxIOopen(xfout, EXIT_ON_FAIL)

   ;  if (get_dim)
      {  const char* fmt
      ;  int format
      ;  long n_cols, n_rows

      ;  if (mclxReadDimensions(xfmx_g, &n_cols, &n_rows))
         mcxDie(1, "mcx q", "reading %s failed", xfmx_g->fn->str)

      ;  format = mclxIOformat(xfmx_g)
      ;  fmt = format == 'b' ? "binary" : format == 'a' ? "interchange" : "?"
      ;  fprintf
         (  xfout->fp
         ,  "%s format,  row x col dimensions are %ld x %ld\n"
         ,  fmt
         ,  n_rows
         ,  n_cols
         )
   ;  }

      if (get_stat)
      {  mclx* mx = mclxRead(xfmx_g, EXIT_ON_FAIL)
      ;  mclv* sums = mclxColSums(mx, MCL_VECTOR_COMPLETE)
      ;  double total = mclvSum(sums)
      ;  dim n_copied = 0

      ;  unsigned long noe = mclxNrofEntries(mx)
      ;  double val_avg, val_sdv, val_med
      ;  double nb_avg , nb_sdv , nb_med
      ;  dim i, j

      ;  pval*  allvals = mcxAlloc(noe * sizeof allvals[0], EXIT_ON_FAIL)
      ;  mclv* colsizes = mclxColSizes(mx, MCL_VECTOR_COMPLETE)

      ;  if (!noe)
         {  fprintf
            (  xfout->fp
            ,  "nb_all=0 nb_avg=0 nb_sdv=0 nb_med=0 val_avg=0 val_sdv=0 val_med=0"
            )
         ;  mcxExit(0)
      ;  }

         val_avg = total / noe
      ;  val_sdv = 0.0

      ;  nb_avg = noe / N_COLS(mx)
      ;  nb_sdv = 0.0
      ;  nb_med = 0.0

      ;  for (i=0;i<N_COLS(mx);i++)
         {  mclv* vec = mx->cols+i
         ;  nb_sdv += (nb_avg - (long) vec->n_ivps) * (nb_avg - (long) vec->n_ivps)
         ;  for (j=0;j<vec->n_ivps;j++)
            {  double val = vec->ivps[j].val
            ;  val_sdv += (val -val_avg) * (val -val_avg)
            ;  allvals[n_copied+j] = val
         ;  }
            n_copied += vec->n_ivps
      ;  }

         val_sdv = sqrt(val_sdv/noe)
      ;  nb_sdv = sqrt(nb_sdv/N_COLS(mx))

      ;  qsort(allvals, noe, sizeof(pval), valcmp)

      ;  if (noe > 1)
         val_med = (allvals[noe/2] + allvals[(noe-1)/2]) / 2.0
      ;  else if (noe == 1)
         val_med = allvals[0]
      ;  else
         val_med = 0.0

      ;  mclvSort(colsizes, mclpValCmp)
      ;  if (N_COLS(mx) > 1)
         nb_med = (colsizes->ivps[N_COLS(mx)/2].val + colsizes->ivps[(N_COLS(mx)-1)/2].val) / 2.0
      ;  else if (N_COLS(mx) == 1)
         nb_med = colsizes->ivps[0].val
      ;  else
         nb_med = 0.0

      ;  fprintf
         (  xfout->fp
         ,  "nb_all=%lu"      " nb_avg=%.3f"    " nb_sdv=%.3f"    " nb_med=%.0f"
            " val_avg=%.3f"   " val_sdv=%.3f"   " val_med=%.3f"
            " val_min=%.3f"   " val_max=%.3f"
            " n_cols=%lu"     " n_rows=%lu"
         ,  noe            ,  nb_avg         ,  nb_sdv         ,  nb_med
         ,  val_avg        ,  val_sdv        ,  val_med
         ,  allvals[0]     ,  allvals[noe-1]
         ,  (ulong) N_COLS(mx)
         ,  (ulong) N_ROWS(mx)
         )

      ;  for (i=0;i<=8;i++)
         {  dim o = (noe * i) / 8
         ;  if (o >= noe)
            o = noe-1
         ;  fprintf(xfout->fp, " val_oc%lu=%.3f", (ulong) i, allvals[o])
      ;  }
         fputc('\n', xfout->fp)
   ;  }
      else if (get_stat_list)
      {  mclx* mx = mclxRead(xfmx_g, EXIT_ON_FAIL)
      ;  mclv* sums = mclxColSums(mx, MCL_VECTOR_COMPLETE)
      ;  dim i
      ;  for (i=0;i<sums->n_ivps;i++)
         {  unsigned long ne = mx->cols[i].n_ivps
         ;  double ew = sums->ivps[i].val
         ;  if (ne)
            fprintf(xfout->fp, "%lu\t%lu\t%g\n", (ulong) sums->ivps[i].idx, ne, ew/ne)
         ;  else
            fprintf(xfout->fp, "%lu\t%lu\t-\n", (ulong) sums->ivps[i].idx, ne)
      ;  }
      }
      return 0
;  }


mcxDispHook* mcxDispHookq
(  void
)
   {  static mcxDispHook qEntry
   =  {  "q"
      ,  "q [options]"
      ,  qOptions
      ,  sizeof(qOptions)/sizeof(mcxOptAnchor) - 1

      ,  qArgHandle
      ,  qInit
      ,  qMain

      ,  0
      ,  0
      ,  MCX_DISP_MANUAL
      }
   ;  return &qEntry
;  }


